### PR TITLE
chore: add flaky retry in weekly CI

### PR DESCRIPTION
**Pull request by RomanBredehoft**
Clean the weekly CI : 
- new workflow : scheduled + on demand , with dedicated status history
- re-run flaky issues like in regular PRs (+ report them on slack notifications for weeklys)
- remove slack reports for regular CIs (keep for weekly + push on main/release branches)
- better slack report message (ex: https://zama-ai.slack.com/archives/C02RX6CF3FV/p1715090858656719)

refs https://github.com/zama-ai/concrete-ml-internal/issues/4030
refs https://github.com/zama-ai/concrete-ml-internal/issues/4428
closes https://github.com/zama-ai/concrete-ml-internal/issues/4390

### CODE REVIEW

File: .github/workflows/continuous-integration.yaml [hunk 1]
```diff
@@ -936,7 +952,7 @@ jobs:
     needs: [build-linux]
     runs-on: ubuntu-20.04
     timeout-minutes: 2
-    if: ${{ always() }}
+    if: success() || failure()
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:955]: avoid running this if all previous jobs are skipped

File: .github/workflows/continuous-integration.yaml [hunk 2]
```diff
@@ -774,22 +775,37 @@ jobs:
...
+              exit 1
+            fi
+          fi
 
-      # Run Pytest on all of our tests (except flaky ones) using PyPI's local wheel during the 
-      # release process
-      - name: PyTest (no flaky) with PyPI local wheel of Concrete ML (release)
+      # Run Pytest on all of our tests (except flaky ones) using PyPI's local wheel in the weekly 
+      # or during the release process
+      - name: PyTest (no flaky) with PyPI local wheel of Concrete ML (weekly, release)
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:806]: avoid flaky tests in this one (redundant)

File: .github/workflows/continuous-integration.yaml [hunk 3]
```diff
@@ -774,22 +775,37 @@ jobs:
         id: pytest_weekly
         if: ${{ fromJSON(env.IS_WEEKLY) && steps.conformance.outcome == 'success' && !cancelled() }}
         run: |
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:777]: retry flaky tests if there are some like in regular CIs

File: .github/workflows/continuous-integration.yaml [hunk 4]
```diff
@@ -1080,12 +1096,25 @@ jobs:
...
           make pytest_macOS_for_GitHub
 
+
+  # Only send a report for the following CI:
+  # - when pushing to main
+  # - when pushing to a release branch
+  # - when running weekly tests
   send-report:
-    if: ${{ always() }}
+    if: |
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:1111]: only report weekly CIs or push to main/release branch : to discuss but I feel it's a bit useless to report all regular CIs on slack

File: .github/workflows/continuous-integration.yaml [hunk 5]
```diff
@@ -60,12 +58,6 @@ on:
         type: boolean
         required: false
         default: false
-  
-  schedule:
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:64]: moved to the new `weekly_tests.yaml`

File: .github/workflows/continuous-integration.yaml [hunk 6]
```diff
@@ -82,7 +74,9 @@ env:
   # differentiate the release CI from regular CIs by using 'inputs.event_name', which should be set
   # to "release" by the release workflow
   IS_PR: ${{ github.event_name == 'pull_request' && inputs.event_name != 'release' }}
-  IS_WEEKLY: ${{ github.event_name == 'schedule' || ((github.event_name == 'workflow_dispatch') && (inputs.event_name == 'weekly')) }}
+  # Run the weekly CI if it has been triggered manually by the weekly workflow, meaning 
+  # 'inputs.event_name' is set to "weekly"
+  IS_WEEKLY: ${{ inputs.event_name == 'weekly'}}
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:79]: now weekly CIs are only triggerd by the new `weekly_tests.yaml` CI

File: .github/workflows/continuous-integration.yaml [hunk 7]
```diff
@@ -147,7 +144,7 @@ jobs:
           # Manage instance type
           INSTANCE_TYPE="c5.4xlarge"
           if [[ "${BUILD_TYPE}" == "weekly" ]]; then
-            INSTANCE_TYPE="m6i.metal"
+            INSTANCE_TYPE="m6i.16xlarge"
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:147]: as for the release, `m6i.metal` started to be less available then before

File: .github/workflows/continuous-integration.yaml [hunk 8]
```diff
@@ -710,12 +714,19 @@ jobs:
...
             )
             || fromJSON(env.IS_WORKFLOW_DISPATCH)
+            || fromJSON(env.IS_WEEKLY)
           )
           && steps.conformance.outcome == 'success' 
           && !cancelled()
         shell: bash +e {0}
         run: |
-          make pytest_and_report
+          if [[ "${{ env.IS_WEEKLY }}" == "true" ]]; then
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:723]: run weekly tests in weekly CIs, but keep the same `flaky re-run` system

File: .github/workflows/continuous-integration.yaml [hunk 9]
```diff
@@ -951,20 +961,8 @@ jobs:
             exit 1
           fi
 
-      - name: Slack Notification
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:954]: keep a single slack notif at the end of the whole CI

File: .github/workflows/continuous-integration.yaml [hunk 10]
```diff
@@ -1080,30 +1081,59 @@ jobs:
       # macOS builds are already long, so we decide not to use --weekly on them, it could be
       # changed. Remark also that, for mac, due to unexpected issues with GitHub, we have a
       # slightly different way to launch pytest
+      # Add support for re-running flaky tests on macOS (intel) CI 
+      # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4428
       - name: PyTest Source Code
         run: |
           make pytest_macOS_for_GitHub
 
+  decide-slack-report:
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:1090]: only report on slaky : 
- weekly CIs
- push to main/release branch CIs 

we can discuss about it but I feel it's a bit useless to report all regular CIs on slack as it makes things very verbose and not very easy to follow

File: .github/workflows/continuous-integration.yaml [hunk 11]
```diff
@@ -1112,20 +1142,65 @@ jobs:
...
+          SLACK_BODY="Build status ([Action URL](${{ env.ACTION_RUN_URL }})):\n\
+            - Linux: ${{ needs.build-linux.result }}\n\n\
+            - macOS (intel): ${{ needs.build-macos-intel.result }}\n\n"
+
+          LINUX_PYTHON_VERSIONS="${{ needs.matrix-preparation.outputs.linux-python-versions }}"
+
+          for linux_python_version in ${LINUX_PYTHON_VERSIONS}; do
+            file_name="failed_tests_slack_list_${linux_python_version}.txt"
+
+            if [ -f ${file_name} ]; then
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:1186]: print flaky tests that got re-run if they were some in the slack notification

File: .github/workflows/weekly_tests.yaml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+name: Weekly Tests
```
Comment (RomanBredehoft) [inline .github/workflows/weekly_tests.yaml:1]: new workflow that only handles weekly CIs, which run:
- each sunday night
- on demand using GH's interface

File: .github/workflows/continuous-integration.yaml [hunk 13]
```diff
@@ -724,7 +735,8 @@ jobs:
           	poetry run python ./script/actions_utils/pytest_failed_test_report.py \
             --pytest-input-report "pytest_report.json" \
             --failed-tests-report "failed_tests_report.json" \
-            --failed-tests-comment "failed_tests_comment.txt"
+            --failed-tests-comment "failed_tests_comment_${{ matrix.python_version }}.txt"
+            --failed-tests-list "failed_tests_slack_list_${{ matrix.python_version }}.txt"
```
Comment (RomanBredehoft) [inline .github/workflows/continuous-integration.yaml:739]: write the list of failed tests that will later be used for reporting in slack flaky tests that got re-run within a weekly CI

**Review by andrei-stoian-zama — APPROVED** [review]

**Pull request merged**